// node-ytmusic-api@1.0.8/dist/YTMusic.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const tough_cookie_1 = require("tough-cookie");
const constants_1 = require("./constants");
const AlbumParser_1 = __importDefault(require("./parsers/AlbumParser"));
const ArtistParser_1 = __importDefault(require("./parsers/ArtistParser"));
const Parser_1 = __importDefault(require("./parsers/Parser"));
const PlaylistParser_1 = __importDefault(require("./parsers/PlaylistParser"));
const SearchParser_1 = __importDefault(require("./parsers/SearchParser"));
const SongParser_1 = __importDefault(require("./parsers/SongParser"));
const VideoParser_1 = __importDefault(require("./parsers/VideoParser"));
const traverse_1 = require("./utils/traverse");
class YTMusic {
    cookiejar;
    config;
    client;
    /**
     * Creates an instance of YTMusic
     * Make sure to call initialize()
     */
    constructor() {
        this.cookiejar = new tough_cookie_1.CookieJar();
        this.config = {};
        this.client = axios_1.default.create({
            baseURL: "https://music.youtube.com/",
            headers: {
                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36",
                "Accept-Language": "en-US,en;q=0.5",
            },
            withCredentials: true,
        });
        this.client.interceptors.request.use(req => {
            if (!req.baseURL)
                return;
            const cookieString = this.cookiejar.getCookieStringSync(req.baseURL);
            if (cookieString) {
                if (!req.headers) {
                    req.headers = {};
                }
                req.headers["Cookie"] = cookieString;
            }
            return req;
        });
        this.client.interceptors.response.use(res => {
            if ("set-cookie" in res.headers) {
                if (!res.config.baseURL)
                    return;
                const setCookie = res.headers["set-cookie"];
                for (const cookieString of [setCookie].flat()) {
                    const cookie = tough_cookie_1.Cookie.parse(`${cookieString}`);
                    if (!cookie)
                        return;
                    this.cookiejar.setCookieSync(cookie, res.config.baseURL);
                }
            }
            return res;
        });
    }
    /**
     * Initializes the API
     */
    async initialize(options) {
        const { cookies, GL, HL } = options ?? {};
        if (cookies) {
            for (const cookieString of cookies.split("; ")) {
                const cookie = tough_cookie_1.Cookie.parse(`${cookieString}`);
                if (!cookie)
                    return;
                this.cookiejar.setCookieSync(cookie, "https://music.youtube.com/");
            }
        }
        const html = (await this.client.get("/")).data;
        const setConfigs = html.match(/ytcfg\.set\(.*\)/) || [];
        const configs = setConfigs
            .map(c => c.slice(10, -1))
            .map(s => {
            try {
                return JSON.parse(s);
            }
            catch {
                return null;
            }
        })
            .filter(j => !!j);
        for (const config of configs) {
            this.config = {
                ...this.config,
                ...config,
            };
        }
        if (!this.config) {
            this.config = {};
        }
        if (GL)
            this.config.GL = GL;
        if (HL)
            this.config.HL = HL;
        return this;
    }
    /**
     * Constructs a basic YouTube Music API request with all essential headers
     * and body parameters needed to make the API work
     *
     * @param endpoint Endpoint for the request
     * @param body Body
     * @param query Search params
     * @returns Raw response from YouTube Music API which needs to be parsed
     */
    async constructRequest(endpoint, body = {}, query = {}) {
        if (!this.config) {
            throw new Error("API not initialized. Make sure to call the initialize() method first");
        }
        this.config = {
            CLIENT_CANARY_STATE: "none",
            DEVICE: "cbr=Chrome&cbrand=apple&cbrver=81.0.4044.129&ceng=WebKit&cengver=537.36&cos=Macintosh&cosver=10_15_4&cplatform=DESKTOP",
            ELEMENT_POOL_DEFAULT_CAP: 0,
            EVENT_ID: "UAvSZdKoL9G96dsP19Sz8A0",
            EXPERIMENT_FLAGS: {
                H5_enable_full_pacf_logging: true,
                H5_use_async_logging: true,
                ab_det_apb_b: true,
                ab_det_el_h: true,
                ab_det_fet_wr: true,
                ab_det_fet_wr_en: true,
                ab_det_gen_re: true,
                action_companion_center_align_description: true,
                allow_skip_networkless: true,
                att_web_record_metrics: true,
                clear_user_partitioned_ls: true,
                compress_gel: true,
                csi_config_handling_infra: true,
                csi_on_gel: true,
                deprecate_csi_has_info: true,
                disable_av_switcher_when_audio_only: true,
                disable_child_node_auto_formatted_strings: true,
                disable_pacf_logging_for_memory_limited_tv: true,
                disable_simple_mixed_direction_formatted_strings: true,
                disable_thumbnail_preloading: true,
                embeds_transport_use_scheduler: true,
                enable_ab_report_on_errorscreen: true,
                enable_ab_rp_int: true,
                enable_cast_on_music_web: true,
                enable_client_sli_logging: true,
                enable_client_streamz_web: true,
                enable_dsa_one_click_ata_translators_infeed_elements: true,
                enable_eom_webview_header: true,
                enable_flow_logging_p4e: true,
                enable_get_account_switcher_endpoint_on_webfe: true,
                enable_ghost_view_rendering_limited_h5_client: true,
                enable_google_payment_billing_command_client_support: true,
                enable_handles_account_menu_switcher: true,
                enable_high_frequency_cookie_rotation: true,
                enable_identity_attributes_for_live_chat: true,
                enable_loggingcontext_trackingparams: true,
                enable_memberships_and_purchases: true,
                enable_mixed_direction_formatted_strings: true,
                enable_modular_player_page_server: true,
                enable_music_css_mss: true,
                enable_music_for_supex: true,
                enable_music_wiz_interop_shims: true,
                enable_music_wiz_next_idom_logging: true,
                enable_names_handles_account_switcher: true,
                enable_pacf_slot_asde_infeed_h5: true,
                enable_pacf_slot_asde_player_byte_h5: true,
                enable_pacf_slot_asde_player_byte_h5_TV: true,
                enable_pass_sdc_get_accounts_list: true,
                enable_pl_r_c: true,
                enable_pl_r_c_s: true,
                enable_pl_r_si_fa: true,
                enable_polymer_resin: true,
                enable_populate_att_psd_in_abe_feedback: true,
                enable_populate_psd_in_abe_feedback: true,
                enable_premium_voluntary_pause: true,
                enable_respect_lang_dir: true,
                enable_sdf_midroll_postroll_player_bytes_video_h5: true,
                enable_sdf_preroll_player_bytes_video_h5: true,
                enable_sdf_preroll_player_bytes_video_tv: true,
                enable_skip_ad_guidance_prompt: true,
                enable_skippable_ads_for_unplugged_ad_pod: true,
                enable_smearing_expansion_dai: true,
                enable_sponsored_ad_badge_on_ytm: true,
                enable_third_party_info: true,
                enable_web_gpay_command_spinner: true,
                enable_web_media_session_metadata_fix: true,
                enable_web_tiered_gel: true,
                enable_web_upload_support: true,
                enable_window_constrained_buy_flow_dialog: true,
                enable_ypc_spinners: true,
                enable_yt_ata_iframe_authuser: true,
                err_on_pl_r_c: true,
                export_networkless_options: true,
                fetch_bid_for_dclk_status: true,
                fill_single_video_with_notify_to_lasr: true,
                gcf_config_store_enabled: true,
                gcf_music_innertube: true,
                graft_gum_response_to_logging_screen: true,
                h5_companion_enable_adcpn_macro_substitution_for_click_pings: true,
                h5_enable_generic_error_logging_event: true,
                h5_inplayer_enable_adcpn_macro_substitution_for_click_pings: true,
                h5_reset_cache_and_filter_before_update_masthead: true,
                hide_endpoint_overflow_on_ytd_display_ad_renderer: true,
                html5_enable_ads_client_monitoring_log_tv: true,
                html5_enable_single_video_vod_ivar_on_pacf: true,
                html5_log_trigger_events_with_debug_data: true,
                il_use_view_model_logging_context: true,
                json_condensed_response: true,
                kev_adb_pg: true,
                kevlar_dropdown_fix: true,
                kevlar_gel_error_routing: true,
                kevlar_response_processor_allowlist_killswitch: true,
                kevlar_service_command_check: true,
                kevlar_toast_manager: true,
                kevlar_use_vimio_behavior: true,
                kevlar_woffle_fallback_image: true,
                live_chat_enable_controller_extraction: true,
                live_chat_enable_rta_manager: true,
                log_errors_through_nwl_on_retry: true,
                log_gel_compression_latency: true,
                log_heartbeat_with_lifecycles: true,
                log_web_endpoint_to_layer: true,
                main_app_controller_extraction_batch_5: true,
                migrate_events_to_ts: true,
                migrate_remaining_web_ad_badges_to_innertube: true,
                music_controller_extraction_batch_1: true,
                music_enable_add_to_save_rename: true,
                music_enable_explore_tab_on_web: true,
                music_enable_improve_your_recommendations_setting: true,
                music_enable_multi_select: true,
                music_enable_radio_steering_in_queues: true,
                music_enable_responsive_radio: true,
                music_enable_responsive_radio_client: true,
                music_enable_responsive_radio_toast_message: true,
                music_enable_save_queue_to_playlist: true,
                music_enable_single_song_queue: true,
                music_enable_sticky_playlist_saving: true,
                music_web_canary_holdback: true,
                music_web_cast_enable_loop: true,
                music_web_cast_set_initial_state: true,
                music_web_delay_watch_next_processing: true,
                music_web_display_av_switcher: true,
                music_web_enable_av_switcher: true,
                music_web_enable_bauhaus_sidenav: true,
                music_web_enable_bauhaus_style_buttons_and_chips: true,
                music_web_enable_bauhaus_style_carousel_buttons: true,
                music_web_enable_capri_redesign: true,
                music_web_enable_client_side_playback_screens: true,
                music_web_enable_drag_drop_upload: true,
                music_web_enable_intent_header: true,
                music_web_enable_new_icon_set: true,
                music_web_enable_persistent_content_warning_flow: true,
                music_web_enable_player_bar_ve_logging_fixes: true,
                music_web_enable_player_page_transition_scroll_persistence: true,
                music_web_enable_reuse_player_queue_item_component: true,
                music_web_enable_select_autoplay_item: true,
                music_web_enable_service_worker: true,
                music_web_enable_wug_subscribe_endpoint: true,
                music_web_fix_queue_logging_race_condition: true,
                music_web_is_canary: true,
                music_web_is_canary_control: true,
                music_web_mark_root_visible: true,
                music_web_ping_ad_start: true,
                music_web_player_context_config: true,
                music_web_prebuffer_autoplay_items: true,
                music_web_respect_playback_content_mode: true,
                music_web_show_history_icon_button: true,
                music_web_volume_slider_granularity: true,
                mweb_deprecate_skip_ve_logging: true,
                mweb_enable_keto_batch_16_2: true,
                mweb_enable_keto_batch_17_2: true,
                networkless_logging: true,
                new_csn_storage_design: true,
                nwl_send_from_memory_when_online: true,
                pageid_as_header_web: true,
                polymer_bad_build_labels: true,
                polymer_verifiy_app_state: true,
                qoe_send_and_write: true,
                replace_closure_window_with_updated_ytwindow_in_studio: true,
                scheduler_use_raf_by_default: true,
                shared_enable_controller_extraction: true,
                shell_load_gcf: true,
                skip_invalid_ytcsi_ticks: true,
                skip_ls_gel_retry: true,
                skip_setting_info_in_csi_data_object: true,
                smarter_ve_dedupping: true,
                st_skip_debug_params: true,
                start_client_gcf: true,
                start_client_gcf_for_player: true,
                start_sending_config_hash: true,
                suppress_error_204_logging: true,
                transport_use_scheduler: true,
                trigger_impression_pings_on_view_search_desktop: true,
                update_log_event_config: true,
                use_core_sm: true,
                use_csi_stp_handler: true,
                use_infogel_early_logging: true,
                use_new_cml: true,
                use_new_in_memory_storage: true,
                use_new_nwl_initialization: true,
                use_new_nwl_saw: true,
                use_new_nwl_stw: true,
                use_new_nwl_wts: true,
                use_player_abuse_bg_library: true,
                use_request_time_ms_header: true,
                use_session_based_sampling: true,
                use_ts_visibilitylogger: true,
                vss_final_ping_send_and_write: true,
                vss_playback_use_send_and_write: true,
                web_api_url: true,
                web_button_rework: true,
                web_button_rework_with_live: true,
                web_csi_action_sampling_enabled: true,
                web_dedupe_ve_grafting: true,
                web_enable_ab_em_rsp: true,
                web_enable_ab_rsp_cl: true,
                web_enable_abd_ref: true,
                web_enable_error_204: true,
                web_enable_voz_audio_feedback: true,
                web_gel_timeout_cap: true,
                web_logging_max_batch_hard_limit: true,
                web_modern_buttons: true,
                web_one_platform_error_handling: true,
                web_playback_associated_ve: true,
                web_prefetch_preload_video: true,
                web_rendererstamper_event_listener: true,
                web_scheduler_auto_init: true,
                web_use_cache_for_image_fallback: true,
                webfe_disable_ab_em_plb: true,
                wiz_use_generic_logging_infra: true,
                woffle_used_state_report: true,
                yt_img_shadow_trigger_show_on_visible: true,
                H5_async_logging_delay_ms: 30000,
                addto_ajax_log_warning_fraction: 0.1,
                log_window_onerror_fraction: 0.1,
                polymer_property_access_logging_percent: 0,
                tv_pacf_logging_sample_rate: 0.01,
                ytidb_transaction_ended_event_rate_limit: 0.02,
                ytidb_transaction_ended_event_rate_limit_session: 0.2,
                ytidb_transaction_ended_event_rate_limit_transaction: 0.1,
                botguard_async_snapshot_timeout_ms: 3000,
                check_navigator_accuracy_timeout_ms: 0,
                client_streamz_web_flush_count: 2,
                client_streamz_web_flush_interval_seconds: 60,
                compression_disable_point: 10,
                compression_performance_threshold: 250,
                gel_min_batch_size: 3,
                gel_queue_timeout_max_ms: 60000,
                hide_cta_for_home_web_video_ads_animate_in_time: 2,
                initial_gel_batch_timeout: 2000,
                max_body_size_to_compress: 500000,
                max_prefetch_window_sec_for_livestream_optimization: 10,
                min_prefetch_offset_sec_for_livestream_optimization: 20,
                music_web_canary_stage: 1,
                music_web_delay_watch_next_ms: 500,
                music_web_list_continuation_prescan_height_px: 3000,
                music_web_session_check_interval_millis: 120000,
                music_web_sidenav_ttl_ms: 14400000,
                network_polling_interval: 30000,
                send_config_hash_timer: 0,
                slow_compressions_before_abandon_count: 4,
                web_foreground_heartbeat_interval_ms: 28000,
                web_gel_debounce_ms: 20000,
                web_logging_max_batch: 400,
                web_smoothness_test_duration_ms: 0,
                web_smoothness_test_method: 0,
                wil_icon_max_concurrent_fetches: 9999,
                ytidb_remake_db_retries: 3,
                ytidb_reopen_db_retries: 3,
                debug_forced_promo_id: "",
                il_payload_scraping: "",
                music_web_body_line_height: "1.4",
                music_web_gpst_url_string: "",
                music_web_title_line_height: "1.2",
                user_preference_collection_initial_browse_id: "FEmusic_tastebuilder",
                web_client_version_override: "",
                kevlar_command_handler_command_banlist: [],
                web_op_signal_type_banlist: [],
            },
            GAPI_HINT_PARAMS: "m;/_/scs/abc-static/_/js/k=gapi.gapi.en.GsbA68hXs80.O/d=1/rs=AHpOoo899t-H8Lxb3OqzMDuPn6TV_i36ag/m=__features__",
            GAPI_HOST: "https://apis.google.com",
            GAPI_LOCALE: "en_US",
            GL: "PL",
            HL: "en",
            HTML_DIR: "ltr",
            HTML_LANG: "en",
            INNERTUBE_API_KEY: "AIzaSyC9XL3ZjWddXya6X74dJoCTL-WEYFDNX30",
            INNERTUBE_API_VERSION: "v1",
            INNERTUBE_CLIENT_NAME: "WEB_REMIX",
            INNERTUBE_CLIENT_VERSION: "1.20240214.01.00",
            INNERTUBE_CONTEXT: {
                client: {
                    hl: "en",
                    gl: "PL",
                    remoteHost: "46.205.197.238",
                    deviceMake: "Apple",
                    deviceModel: "",
                    visitorData: "CgtmTTVuSkd5aF9lNCjQlsiuBjIOCgJQTBIIEgQSAgsMIG4%3D",
                    userAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.129 Safari/537.36,gzip(gfe)",
                    clientName: "WEB_REMIX",
                    clientVersion: "1.20240214.01.00",
                    osName: "Macintosh",
                    osVersion: "10_15_4",
                    originalUrl: "https://music.youtube.com/?cbrd=1&ucbcb=1",
                    platform: "DESKTOP",
                    clientFormFactor: "UNKNOWN_FORM_FACTOR",
                    configInfo: [Object],
                    browserName: "Chrome",
                    browserVersion: "81.0.4044.129",
                    acceptHeader: "application/json, text/plain, */*",
                    deviceExperimentId: "ChxOek16Tmprek9URTRNVGMxTURZeU5URTJOdz09ENCWyK4GGNCWyK4G",
                },
                user: { lockedSafetyMode: false },
                request: { useSsl: true },
                clickTracking: { clickTrackingParams: "IhMI0pClqYS1hAMV0V56BR1X6gze" },
            },
            INNERTUBE_CONTEXT_CLIENT_NAME: 67,
            INNERTUBE_CONTEXT_CLIENT_VERSION: "1.20240214.01.00",
            INNERTUBE_CONTEXT_GL: "PL",
            INNERTUBE_CONTEXT_HL: "en",
            LATEST_ECATCHER_SERVICE_TRACKING_PARAMS: { "client.name": "WEB_REMIX" },
            LOGGED_IN: false,
            PAGE_BUILD_LABEL: "youtube.music.web.client_20240214_01_RC00",
            PAGE_CL: 606892047,
            scheduler: { useRaf: true, timeout: 20 },
            SERVER_NAME: "WebFE",
            SIGNIN_URL: "https://accounts.google.com/ServiceLogin?ltmpl=music&service=youtube&uilel=3&passive=true&continue=https%3A%2F%2Fwww.youtube.com%2Fsignin%3Faction_handle_signin%3Dtrue%26app%3Ddesktop%26hl%3Den%26next%3Dhttps%253A%252F%252Fmusic.youtube.com%252F%253Fcbrd%253D1%2526ucbcb%253D1%26feature%3D__FEATURE__&hl=en",
            WEB_PLAYER_CONTEXT_CONFIGS: {
                WEB_PLAYER_CONTEXT_CONFIG_ID_MUSIC_WATCH: {
                    rootElementId: "movie_player",
                    jsUrl: "/s/player/5683fc5e/player_ias.vflset/en_US/base.js",
                    cssUrl: "/s/player/5683fc5e/www-player.css",
                    contextId: "WEB_PLAYER_CONTEXT_CONFIG_ID_MUSIC_WATCH",
                    eventLabel: "detailpage",
                    contentRegion: "PL",
                    hl: "en_US",
                    hostLanguage: "en",
                    innertubeApiKey: "AIzaSyC9XL3ZjWddXya6X74dJoCTL-WEYFDNX30",
                    innertubeApiVersion: "v1",
                    innertubeContextClientVersion: "1.20240214.01.00",
                    controlsType: 0,
                    disableKeyboardControls: true,
                    disableRelatedVideos: true,
                    annotationsLoadPolicy: 3,
                    device: [Object],
                    serializedExperimentIds: "23983296,24004644,24007246,24080738,24135310,24362604,24362631,24385728,24439361,24524098,24560416,24566687,51006181,51010235,51017346,51026715,51037540,51053689,51063643,51085358,51091331,51092929,51096389,51098297,51098299,51099452,51102317,51103044,51104002,51105868,51106995,51109696,51116067,51119400,51119911",
                    disableSharing: true,
                    hideInfo: true,
                    disableWatchLater: true,
                    mobileIphoneSupportsInlinePlayback: true,
                    isMobileDevice: false,
                    cspNonce: "FKHPD8mfeLs-DzWoVsBcQA",
                    canaryState: "none",
                    enableCsiLogging: true,
                    datasyncId: "Va38a1e8e||",
                    allowWoffleManagement: true,
                    canaryStage: "",
                },
            },
            XSRF_FIELD_NAME: "session_token",
            XSRF_TOKEN: "QUFFLUhqbjlOU2xNREhYN3FwZDdLZXdzREZZWDdPdTJxd3xBQ3Jtc0trQTZ1aU5JOXB3ZDhkekpTNHhGVGRDNGtaNHZnaFNyNjNaVkRjY2RIbnhtNnRQbzFlU0NnTkZhdmN6Y3g3c1U2U0NQckZod1ZuUnBjNmZ5REcyVF9sb2UydVpQanVzS3NxSENvWkFFdnUwV1ZaNEhoTQ==",
            YPC_MB_URL: "https://payments.youtube.com/payments/v4/js/integrator.js?ss=md",
            YTR_FAMILY_CREATION_URL: "https://families.google.com/webcreation?usegapi=1",
            SERVER_VERSION: "prod",
            LOCALE: "en_US",
            REUSE_COMPONENTS: true,
            STAMPER_STABLE_LIST: false,
            DATASYNC_ID: "Va38a1e8e||",
            SERIALIZED_CLIENT_CONFIG_DATA: "CNCWyK4GEN3o_hIQ_bawBRComrAFELfvrwUQiIewBRDr6P4SEKmN_xIQyLiwBRC--a8FEMn3rwUQpoGwBRDViLAFEPywsAUQ9quwBRDnuq8FEMq3sAUQmrCwBRC-irAFEJrwrwUQvPmvBRCst68FEP24_RIQ6sOvBRD0q7AFELfq_hIQ4riwBRDh8q8FELiqsAUQt6uwBRDZya8FELiLrgUQ3bCwBRClkLAFEPCcsAUQopKwBRDerrAFEL2ZsAUQvbauBRCu1P4SEKXC_hIQzN-uBRCwsrAFEM-osAUQqfevBRCcj_8SEKKBsAUQ7rOwBRCQsrAFEPqnsAUQ_IWwBRDmsrAFEKKusAUQiOOvBRDQjbAFEPOhsAUQ0-GvBRDMrv4SEI27sAU%3D",
            MDX_CONFIG: {
                device: "Desktop",
                app: "youtube.m-desktop",
                appId: "0F5096E8",
                disableDial: true,
                theme: "m",
                loadCastApiSetupScript: true,
                capabilities: ["que", "mlm"],
                disableAutomaticScreenCache: true,
                forceMirroring: true,
                enableConnectWithInitialState: true,
            },
            CLIENT_PROTOCOL: "HTTP/1.1",
            CLIENT_TRANSPORT: "tcp",
            EOM_VISITOR_DATA: "CgtmTTVuSkd5aF9lNCjQlsiuBjIOCgJQTBIIEgQSAgsMIG4%3D",
            TIME_CREATED_MS: 1708264272835,
            VALID_SESSION_TEMPDATA_DOMAINS: [
                "youtu.be",
                "youtube.com",
                "www.youtube.com",
                "web-green-qa.youtube.com",
                "web-release-qa.youtube.com",
                "web-integration-qa.youtube.com",
                "m.youtube.com",
                "mweb-green-qa.youtube.com",
                "mweb-release-qa.youtube.com",
                "mweb-integration-qa.youtube.com",
                "studio.youtube.com",
                "studio-green-qa.youtube.com",
                "studio-integration-qa.youtube.com",
            ],
            RAW_COLD_CONFIG_GROUP: {
                configData: "CNCWyK4GEO-6rQUQxIWuBRC9tq4FENjurgUQpMWvBRCyxq8FEL6KsAUQ6YywBRCTlrAFEMCcsAUQ8JywBRDQpbAFENSnsAUQ3KuwBRCirrAFEM2usAUQ3q6wBRCZr7AFEOaysAUQ7rOwBRD9trAFEMq3sAUQyLiwBRDiuLAFEI27sAUQ0ruwBRCKvbAFGjJBT2pGb3gyRWotSmFnelF3OFh5Wm92UjZzbXptdWhvU05PNVhQeGxRamZULVhMZWhsUSIyQU9qRm94MkVqLUphZ3pRdzhYeVpvdlI2c216bXVob1NOTzVYUHhsUWpmVC1YTGVobFEqLENBTVNIQTBOZ3Bhb0FzNFRwaF9lTWhVSmtvTFFESmJkQkpEWkJKX1J6QXM9",
                musicColdConfig: {
                    enableKnightRider: true,
                    enableShortStack: true,
                    enableBottomSheetAccountSwitcher: true,
                    enableShowLibrary: true,
                    enableRemixPlayerPage: false,
                    musicEnableHomePageDiskCaching: false,
                    iosEnableLightweightHomepage: false,
                    enableShowDownloadInLibrary: true,
                    androidEnableRxForPlayerEvents: false,
                    enableMusicDownloadsAutoOffline: false,
                    enableWoffle: false,
                },
            },
            RAW_HOT_CONFIG_GROUP: {
                musicHotConfig: {
                    iosEnableLightweightCollectionview: true,
                    enableOfflineLikedTab: false,
                    enableRestrictedModeSetting: true,
                    enablePlaybackLogging: false,
                    enableAndroidShortcuts: false,
                    prebufferContentLengthMs: 0,
                    prebufferCountdownTimeMs: 0,
                    enableSongOffline: false,
                    enableLoopOnMissingNextEndpoint: false,
                    enableAutoOptInForNotifications: false,
                    prefetchMaxRetries: 0,
                    prefetchRetryIntervalMs: 30000,
                    musicDisplayConfig: [Object],
                    enableNetworkChangeSnackbar: false,
                    enableIosAirplayButton: false,
                    enableMediaBrowserService: false,
                    enableRestorePlaybackState: false,
                    enableWatchHistoryNotifierConditionalRenderer: false,
                    enablePlaybackQueue: true,
                    enableInnertubeSearchSuggestionsService: false,
                    checkMultiwindowBeforeBackground: false,
                    enableQuickseekActions: false,
                    enableMediaKeyActions: false,
                    enableRemixOfflineAlbumDetailPage: true,
                    enableRemixDownloadsSection: false,
                    enableRemixOfflinePlaylistDetailPage: true,
                    musicEnableAmplifierInWatchNextService: true,
                    enableMediaBrowserServiceLogging: false,
                    musicEnableAndroidPersistentQueue: false,
                },
            },
            SERIALIZED_HOT_HASH_DATA: "CNCWyK4GEhMzNDY3MDk1OTAwMzUzNDU5MTE0GNCWyK4GKJTk_BIo3JP9EijGsv0SKJ6R_hIomq3-EijIyv4SKN3O_hIoqOH-Eijh9v4SKKD3_hIop_z-EiiygP8SKMCD_xIon4n_Eiipjf8SKJyP_xIojpH_EiiPkf8SKJKR_xIomJH_Eii5kv8SKLOT_xIomJX_EjIyQU9qRm94MkVqLUphZ3pRdzhYeVpvdlI2c216bXVob1NOTzVYUHhsUWpmVC1YTGVobFE6MkFPakZveDJFai1KYWd6UXc4WHlab3ZSNnNtem11aG9TTk81WFB4bFFqZlQtWExlaGxRQiRDQU1TRmcwUDJJXzVGY29BcURrVkRZM2l6UXlMN2dHdTZBcz0%3D",
            SERIALIZED_COLD_HASH_DATA: "CNCWyK4GEhQxMzk1NDgyMDAzNjEzNDg0OTU0MhjQlsiuBjIyQU9qRm94MkVqLUphZ3pRdzhYeVpvdlI2c216bXVob1NOTzVYUHhsUWpmVC1YTGVobFE6MkFPakZveDJFai1KYWd6UXc4WHlab3ZSNnNtem11aG9TTk81WFB4bFFqZlQtWExlaGxRQixDQU1TSEEwTmdwYW9BczRUcGhfZU1oVUprb0xRREpiZEJKRFpCSl9SekFzPQ%3D%3D",
            USE_EMBEDDED_INNERTUBE_DATA: true,
            VISIBILITY_ROOT: "ytmusic-app",
            YTMUSIC_ICON_SRC: "//music.youtube.com/img/on_platform_logo.svg",
            YTMUSIC_LOGO_SRC: "//music.youtube.com/img/on_platform_logo_dark.svg",
            UPLOAD_URL: "https://upload.youtube.com/upload/usermusic/http",
            TRANSFER_PAGE_SIGNIN_URL: "https://accounts.google.com/ServiceLogin?ltmpl=music&service=youtube&uilel=3&passive=true&continue=https%3A%2F%2Fwww.youtube.com%2Fsignin%3Faction_handle_signin%3Dtrue%26app%3Ddesktop%26hl%3Den%26next%3Dhttps%253A%252F%252Fmusic.youtube.com%252Ftransfer%26feature%3D__FEATURE__&hl=en",
            LOGOUT_URL: "/logout",
            IS_SUBSCRIBER: false,
            IS_MOBILE_WEB: false,
            INITIAL_ENDPOINT: '{"clickTrackingParams":"IhMIiammqYS1hAMV7kB6BR0hHgEZMghleHRlcm5hbA==","browseEndpoint":{"browseId":"FEmusic_home"}}',
            HOTKEY_DIALOG: {
                title: { runs: [Array] },
                sections: [[Object], [Object], [Object]],
            },
            DEFAULT_ALBUM_IMAGE_SRC: "//music.youtube.com/img/default_album_image.svg",
            AUDIO_QUALITY: "AUDIO_QUALITY_MEDIUM",
            ADD_SCRAPER_ATTRIBUTES: false,
            ACTIVE_ACCOUNT_IS_MADISON_ACCOUNT: false,
            YTMUSIC_WHITE_ICON_SRC: "//music.youtube.com/img/white_icon.svg",
            YTMUSIC_WHITE_LOGO_SRC: "//music.youtube.com/img/white_logo.svg",
            ACTIVE_ACCOUNT_CAN_UPLOAD: false,
        };
        const headers = {
            ...this.client.defaults.headers,
            "x-origin": this.client.defaults.baseURL,
            "X-Goog-Visitor-Id": this.config.VISITOR_DATA || "",
            "X-YouTube-Client-Name": this.config.INNERTUBE_CONTEXT_CLIENT_NAME,
            "X-YouTube-Client-Version": this.config.INNERTUBE_CLIENT_VERSION,
            "X-YouTube-Device": this.config.DEVICE,
            "X-YouTube-Page-CL": this.config.PAGE_CL,
            "X-YouTube-Page-Label": this.config.PAGE_BUILD_LABEL,
            "X-YouTube-Utc-Offset": String(-new Date().getTimezoneOffset()),
            "X-YouTube-Time-Zone": new Intl.DateTimeFormat().resolvedOptions().timeZone,
        };
        const searchParams = new URLSearchParams({
            ...query,
            alt: "json",
            key: this.config.INNERTUBE_API_KEY,
        });
        const res = await this.client.post(`youtubei/${this.config.INNERTUBE_API_VERSION}/${endpoint}?${searchParams.toString()}`, {
            context: {
                capabilities: {},
                client: {
                    clientName: this.config.INNERTUBE_CLIENT_NAME,
                    clientVersion: this.config.INNERTUBE_CLIENT_VERSION,
                    experimentIds: [],
                    experimentsToken: "",
                    gl: this.config.GL,
                    hl: this.config.HL,
                    locationInfo: {
                        locationPermissionAuthorizationStatus: "LOCATION_PERMISSION_AUTHORIZATION_STATUS_UNSUPPORTED",
                    },
                    musicAppInfo: {
                        musicActivityMasterSwitch: "MUSIC_ACTIVITY_MASTER_SWITCH_INDETERMINATE",
                        musicLocationMasterSwitch: "MUSIC_LOCATION_MASTER_SWITCH_INDETERMINATE",
                        pwaInstallabilityStatus: "PWA_INSTALLABILITY_STATUS_UNKNOWN",
                    },
                    utcOffsetMinutes: -new Date().getTimezoneOffset(),
                },
                request: {
                    internalExperimentFlags: [
                        {
                            key: "force_music_enable_outertube_tastebuilder_browse",
                            value: "true",
                        },
                        {
                            key: "force_music_enable_outertube_playlist_detail_browse",
                            value: "true",
                        },
                        {
                            key: "force_music_enable_outertube_search_suggestions",
                            value: "true",
                        },
                    ],
                    sessionIndex: {},
                },
                user: {
                    enableSafetyMode: false,
                },
            },
            ...body,
        }, {
            responseType: "json",
            headers,
        });
        return "responseContext" in res.data ? res.data : res;
    }
    /**
     * Get a list of search suggestiong based on the query
     *
     * @param query Query string
     * @returns Search suggestions
     */
    async getSearchSuggestions(query) {
        return (0, traverse_1.traverseList)(await this.constructRequest("music/get_search_suggestions", {
            input: query,
        }), "query");
    }
    /**
     * Searches YouTube Music API for results
     *
     * @param query Query string
     */
    async search(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: null,
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer")
            .map(SearchParser_1.default.parse)
            .filter(Boolean);
    }
    /**
     * Searches YouTube Music API for songs
     *
     * @param query Query string
     */
    async searchSongs(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: "Eg-KAQwIARAAGAAgACgAMABqChAEEAMQCRAFEAo%3D",
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer").map(SongParser_1.default.parseSearchResult);
    }
    /**
     * Searches YouTube Music API for videos
     *
     * @param query Query string
     */
    async searchVideos(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: "Eg-KAQwIABABGAAgACgAMABqChAEEAMQCRAFEAo%3D",
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer").map(VideoParser_1.default.parseSearchResult);
    }
    /**
     * Searches YouTube Music API for artists
     *
     * @param query Query string
     */
    async searchArtists(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: "Eg-KAQwIABAAGAAgASgAMABqChAEEAMQCRAFEAo%3D",
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer").map(ArtistParser_1.default.parseSearchResult);
    }
    /**
     * Searches YouTube Music API for albums
     *
     * @param query Query string
     */
    async searchAlbums(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: "Eg-KAQwIABAAGAEgACgAMABqChAEEAMQCRAFEAo%3D",
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer").map(AlbumParser_1.default.parseSearchResult);
    }
    /**
     * Searches YouTube Music API for playlists
     *
     * @param query Query string
     */
    async searchPlaylists(query) {
        const searchData = await this.constructRequest("search", {
            query,
            params: "Eg-KAQwIABAAGAAgACgBMABqChAEEAMQCRAFEAo%3D",
        });
        return (0, traverse_1.traverseList)(searchData, "musicResponsiveListItemRenderer").map(PlaylistParser_1.default.parseSearchResult);
    }
    /**
     * Get all possible information of a Song
     *
     * @param videoId Video ID
     * @returns Song Data
     */
    async getSong(videoId) {
        if (!videoId.match(/^[a-zA-Z0-9-_]{11}$/))
            throw new Error("Invalid videoId");
        const data = await this.constructRequest("player", { videoId });
        const song = SongParser_1.default.parse(data);
        if (song.videoId !== videoId)
            throw new Error("Invalid videoId");
        return song;
    }
    /**
     * Get all possible information of a Video
     *
     * @param videoId Video ID
     * @returns Video Data
     */
    async getVideo(videoId) {
        if (!videoId.match(/^[a-zA-Z0-9-_]{11}$/))
            throw new Error("Invalid videoId");
        const data = await this.constructRequest("player", { videoId });
        const video = VideoParser_1.default.parse(data);
        if (video.videoId !== videoId)
            throw new Error("Invalid videoId");
        return video;
    }
    /**
     * Get lyrics of a specific Song
     *
     * @param videoId Video ID
     * @returns Lyrics
     */
    async getLyrics(videoId) {
        if (!videoId.match(/^[a-zA-Z0-9-_]{11}$/))
            throw new Error("Invalid videoId");
        const data = await this.constructRequest("next", { videoId });
        const browseId = (0, traverse_1.traverse)((0, traverse_1.traverseList)(data, "tabs", "tabRenderer")[1], "browseId");
        const lyricsData = await this.constructRequest("browse", { browseId });
        const lyrics = (0, traverse_1.traverseString)(lyricsData, "description", "runs", "text");
        return lyrics
            ? lyrics
                .replaceAll("\r", "")
                .split("\n")
                .filter(v => !!v)
            : null;
    }
    /**
     * Get all possible information of an Artist
     *
     * @param artistId Artist ID
     * @returns Artist Data
     */
    async getArtist(artistId) {
        const data = await this.constructRequest("browse", {
            browseId: artistId,
        });
        return ArtistParser_1.default.parse(data, artistId);
    }
    /**
     * Get all of Artist's Songs
     *
     * @param artistId Artist ID
     * @returns Artist's Songs
     */
    async getArtistSongs(artistId) {
        const artistData = await this.constructRequest("browse", { browseId: artistId });
        const browseToken = (0, traverse_1.traverse)(artistData, "musicShelfRenderer", "title", "browseId");
        if (browseToken instanceof Array)
            return [];
        const songsData = await this.constructRequest("browse", { browseId: browseToken });
        const continueToken = (0, traverse_1.traverse)(songsData, "continuation");
        const moreSongsData = await this.constructRequest("browse", {}, { continuation: continueToken });
        return [
            ...(0, traverse_1.traverseList)(songsData, "musicResponsiveListItemRenderer"),
            ...(0, traverse_1.traverseList)(moreSongsData, "musicResponsiveListItemRenderer"),
        ].map(s => SongParser_1.default.parseArtistSong(s, {
            artistId,
            name: (0, traverse_1.traverseString)(artistData, "header", "title", "text"),
        }));
    }
    /**
     * Get all of Artist's Albums
     *
     * @param artistId Artist ID
     * @returns Artist's Albums
     */
    async getArtistAlbums(artistId) {
        const artistData = await this.constructRequest("browse", {
            browseId: artistId,
        });
        const artistAlbumsData = (0, traverse_1.traverseList)(artistData, "musicCarouselShelfRenderer")[0];
        const browseBody = (0, traverse_1.traverse)(artistAlbumsData, "moreContentButton", "browseEndpoint");
        const albumsData = await this.constructRequest("browse", browseBody);
        return (0, traverse_1.traverseList)(albumsData, "musicTwoRowItemRenderer").map(item => AlbumParser_1.default.parseArtistAlbum(item, {
            artistId,
            name: (0, traverse_1.traverseString)(albumsData, "header", "runs", "text"),
        }));
    }
    /**
     * Get all possible information of an Album
     *
     * @param albumId Album ID
     * @returns Album Data
     */
    async getAlbum(albumId) {
        const data = await this.constructRequest("browse", {
            browseId: albumId,
        });
        return AlbumParser_1.default.parse(data, albumId);
    }
    /**
     * Get all possible information of a Playlist except the tracks
     *
     * @param playlistId Playlist ID
     * @returns Playlist Data
     */
    async getPlaylist(playlistId) {
        if (playlistId.startsWith("PL"))
            playlistId = "VL" + playlistId;
        const data = await this.constructRequest("browse", {
            browseId: playlistId,
        });
        return PlaylistParser_1.default.parse(data, playlistId);
    }
    /**
     * Get all videos in a Playlist
     *
     * @param playlistId Playlist ID
     * @returns Playlist's Videos
     */
    async getPlaylistVideos(playlistId) {
        if (playlistId.startsWith("PL"))
            playlistId = "VL" + playlistId;
        const playlistData = await this.constructRequest("browse", {
            browseId: playlistId,
        });
        const songs = (0, traverse_1.traverseList)(playlistData, "musicPlaylistShelfRenderer", "musicResponsiveListItemRenderer");
        let continuation = (0, traverse_1.traverse)(playlistData, "continuation");
        while (!(continuation instanceof Array)) {
            const songsData = await this.constructRequest("browse", {}, { continuation });
            songs.push(...(0, traverse_1.traverseList)(songsData, "musicResponsiveListItemRenderer"));
            continuation = (0, traverse_1.traverse)(songsData, "continuation");
        }
        return songs.map(VideoParser_1.default.parsePlaylistVideo);
    }
    /**
     * Get content for the home page.
     *
     * @returns Mixed HomePageContent
     */
    async getHome() {
        const results = [];
        const page = await this.constructRequest("browse", { browseId: constants_1.FE_MUSIC_HOME });
        (0, traverse_1.traverseList)(page, "sectionListRenderer", "contents").forEach(content => {
            const parsed = Parser_1.default.parseMixedContent(content);
            parsed && results.push(parsed);
        });
        let continuation = (0, traverse_1.traverseString)(page, "continuation");
        while (continuation) {
            const nextPage = await this.constructRequest("browse", {}, { continuation });
            (0, traverse_1.traverseList)(nextPage, "sectionListContinuation", "contents").forEach(content => {
                const parsed = Parser_1.default.parseMixedContent(content);
                parsed && results.push(parsed);
            });
            continuation = (0, traverse_1.traverseString)(nextPage, "continuation");
        }
        return results;
    }
}
exports.default = YTMusic;
